pkg/util/resource/resource_manager.go

```diff
@@ -0,0 +1,297 @@
+	return func(res *wrapper) {
+		res.keepAliveTime = keepAliveTime
+	}
+}
+
+func NewResource(opts ...Option) Resource {
+	w := &wrapper{}
+	for _, opt := range opts {
+		opt(w)
+	}
+	return w
+}
+
+func NewSimpleResource(obj any, typ, name string, keepAliveTime time.Duration, closeFunc func()) Resource {
+	return NewResource(WithObj(obj), WithType(typ), WithName(name), WithKeepAliveTime(keepAliveTime), WithCloseFunc(closeFunc))
+}
+
+type Manager interface {
+	Get(typ, name string, newResourceFunc NewResourceFunc) (Resource, error)
+	Delete(typ, name string) Resource
+	Close()
+}
+
+type item struct {
+	res            Resource
+	updateTimeChan chan int64
+	deleteMark     chan struct{}
+	expiration     int64
+}
+
+type manager struct {
+	resources              map[string]map[string]*item // key: resource type, value: resource name -> resource
+	checkInterval          time.Duration
+	defaultExpiration      time.Duration
+	defaultTypeExpirations map[string]time.Duration // key: resource type, value: expiration
+	mu                     sync.RWMutex
+	stop                   chan struct{}
+	stopOnce               sync.Once
+}
+
+func NewManager(checkInterval, defaultExpiration time.Duration, defaultTypeExpirations map[string]time.Duration) Manager {
+	if checkInterval <= 0 {
+		checkInterval = DefaultCheckInterval
+	}
+	if defaultExpiration <= 0 {
+		defaultExpiration = DefaultExpiration
+	}
+	if defaultTypeExpirations == nil {
+		defaultTypeExpirations = make(map[string]time.Duration)
+	}
+	m := &manager{
+		resources:              make(map[string]map[string]*item),
+		checkInterval:          checkInterval,
+		defaultExpiration:      defaultExpiration,
+		defaultTypeExpirations: defaultTypeExpirations,
+		stop:                   make(chan struct{}),
+	}
+	go m.backgroundGC()
+	return m
+}
+
+func (m *manager) backgroundGC() {
+	ticker := time.NewTicker(m.checkInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ticker.C:
+			m.gc()
+		case <-m.stop:
+			m.mu.Lock()
+			for _, typMap := range m.resources {
+				for _, item := range typMap {
+					item.res.Close()
+				}
+			}
+			m.resources = nil
+			m.mu.Unlock()
+			return
+		}
+	}
+}
+
+func (m *manager) gc() {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	now := time.Now().UnixNano()
+	for typ, typMap := range m.resources {
+		for resName, item := range typMap {
+			select {
+			case lastTime := <-item.updateTimeChan:
+				if item.expiration >= 0 {
+					item.expiration = lastTime
+				}
+			case <-item.deleteMark:
+				item.res.Close()
+				delete(typMap, resName)
+			default:
+				if item.expiration >= 0 && item.expiration <= now {
+					item.res.Close()
+					delete(typMap, resName)
+				}
+			}
+		}
+		if len(typMap) == 0 {
+			delete(m.resources, typ)
+		}
+	}
+}
+
+func (m *manager) updateExpire(item *item) {
+	select {
+	case item.updateTimeChan <- time.Now().UnixNano() + item.res.KeepAliveTime().Nanoseconds():
+	default:
+	}
+}
```

Comment:
using `updateTimeChan` with length = 1 will cause multiple update failure if the buffer is full and gc logic will recycle this item by mistake

Reply:
The expiration time here is guaranteed to be twice the check period, so that errors will not occur. In addition, even if the resource is recycled by mistake, the cost is not high.